File: it-agents/skills/code-architect/references/code-examples/go/factory_method.go
// Factory Method: a creator method returns products via an interface,
// letting subclasses decide which concrete product to build.
package main

import "fmt"

// Product is the common interface returned by every factory.
type Product interface {
	Operate() string
}

type alphaProduct struct{}

func (alphaProduct) Operate() string { return "alpha behaviour" }

type betaProduct struct{}

func (betaProduct) Operate() string { return "beta behaviour" }

// Creator declares the factory method; concrete creators override it.
type Creator interface {
	Make() Product
}

type alphaCreator struct{}

func (alphaCreator) Make() Product { return alphaProduct{} }

type betaCreator struct{}

func (betaCreator) Make() Product { return betaProduct{} }

func describe(c Creator) string {
	return c.Make().Operate()
}

func main() {
	creators := []Creator{alphaCreator{}, betaCreator{}}
	for _, c := range creators {
		fmt.Println(describe(c))
	}
}

File: it-agents/skills/code-architect/references/code-examples/rust/factory_method.rs
// Factory Method — a creator returns products through a trait object, letting
// subclasses (here, concrete creators) decide which concrete product to build.

trait Product {
    fn describe(&self) -> String;
}

struct AlphaProduct;
struct BetaProduct;

impl Product for AlphaProduct {
    fn describe(&self) -> String {
        "Alpha product (lightweight)".to_string()
    }
}

impl Product for BetaProduct {
    fn describe(&self) -> String {
        "Beta product (heavy duty)".to_string()
    }
}

// The Creator declares the factory method; default logic uses whatever it returns.
trait Creator {
    fn create(&self) -> Box<dyn Product>;

    fn produce_and_report(&self) -> String {
        let product = self.create();
        format!("Creator delivered -> {}", product.describe())
    }
}

struct AlphaCreator;
struct BetaCreator;

impl Creator for AlphaCreator {
    fn create(&self) -> Box<dyn Product> {
        Box::new(AlphaProduct)
    }
}

impl Creator for BetaCreator {
    fn create(&self) -> Box<dyn Product> {
        Box::new(BetaProduct)
    }
}

fn main() {
    let creators: Vec<Box<dyn Creator>> = vec![Box::new(AlphaCreator), Box::new(BetaCreator)];
    for creator in &creators {
        println!("{}", creator.produce_and_report());
    }
}
